### ISSUE TITLE
Reusing Negroni Middleware (Avoiding Redundancy)

### ISSUE THREAD

**Issue by williamhaley** [urfave/negroni#167]
I have a lot of different routes that share the same "base" middleware.  For example's sake, let's say that I re-use the same three negroni middlewares in a bunch of different route groups.

```
LogMiddleware
BaseMiddleware1
BaseMiddleware2
```

I can do something like this.

```
router.PathPrefix("/api/v1/users").Handler(negroni.New(
	&LogMiddleware{},
	&BaseMiddleware1{},
	&BaseMiddleware2{},
	&ApiSpecificMiddleware{},
	&UsersSpecificMiddleware{},
	negroni.Wrap(usersRouter),
))
router.PathPrefix("/api/v1/posts").Handler(negroni.New(
	&LogMiddleware{},
	&BaseMiddleware1{},
	&BaseMiddleware2{},
	&ApiSpecificMiddleware{},
	&PostsSpecificMiddleware{},
	negroni.Wrap(postsRouter),
))
...
```

That works perfectly well.

```
2017/01/04 14:02:55 LOGGER Middleware: Received request: GET /api/v1/posts/all
2017/01/04 14:02:55 Base Middleware 1
2017/01/04 14:02:55 Base Middleware 2
2017/01/04 14:02:55 ApiSpecificMiddleware
2017/01/04 14:02:55 PostsSpecificMiddleware
2017/01/04 14:02:55 PostsHandler
2017/01/04 14:02:55 LOGGER Middleware: Request complete: 200 OK
```

But I feel that's a bit wordy. There's enough code repetition that I feel like there should be a better solution. Especially if I add more middleware in the future.

I was hoping instead I could avoid duplication and do something like this.

```
// Negroni middleware that should be shared between multiple route groups.
baseNegroniMiddleware := negroni.New(
	&LogMiddleware{},
	&BaseMiddleware1{},
	&BaseMiddleware2{},
)
...
router.PathPrefix("/api/v1/posts").Handler(negroni.New(
	negroni.Wrap(baseNegroniMiddleware),
	&ApiSpecificMiddleware{},
	&UsersSpecificMiddleware{},
	negroni.Wrap(usersRouter),
))
```

That _sort of_ works, but not really.

```
2017/01/04 14:03:00 LOGGER Middleware: Received request: GET /api/v1/posts/all
2017/01/04 14:03:00 Base Middleware 1
2017/01/04 14:03:00 Base Middleware 2
2017/01/04 14:03:00 LOGGER Middleware: Request complete: 0
2017/01/04 14:03:00 ApiSpecificMiddleware
2017/01/04 14:03:00 PostsSpecificMiddleware
2017/01/04 14:03:00 PostsHandler
```

Note that things end up out of order. My logger middleware breaks and does not fire in the right order.

Here's what the LogMiddleware is doing, which is pretty similar to the [negroni logger](https://github.com/urfave/negroni/blob/master/logger.go).

```
func (m *LogMiddleware) ServeHTTP(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
	log.Printf("LOGGER Middleware: Received request: %s %s ", r.Method, r.URL.Path)
	next(rw, r)
	rsp := rw.(negroni.ResponseWriter)
	log.Printf("LOGGER Middleware: Request complete: %v %s", rsp.Status(), http.StatusText(rsp.Status()))
}
```

Is there some syntax or pattern for reuse of negroni middleware like I'm attempting? Is what I'm doing sensible, but maybe I'm just making a mistake?

**Comment by yanfali** [comment]
Have you looked at alice for route specific middleware? I tend to use negroni for cross cutting middleware -- static files serving, authentication and authorization -- and alice for route specific stuff.

https://github.com/justinas/alice

**Comment by williamhaley** [comment]
Thanks for the suggestion @yanfali, but I'm not sure Alice would fit my needs. I'm leveraging this sort of power in Negroni, which I believe I would lose with Alice unless I rethought much of my middleware implementation.

```
func MyMiddleware(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
  // do some stuff before
  next(rw, r)
  // do some stuff after
}
```

Either way, I'd rather stick with negroni (since I already have multiple apps in my ecosystem that are exclusively designed to use negroni middleware syntax) if I can help it.

Another possibility I just realized is using an array of negroni handlers and managing it myself. It's rather ugly, but it works.

```
baseMiddleware := []negroni.Handler{
	&LogMiddleware{},
	&BaseMiddleware1{},
	&BaseMiddleware2{},
}
apiMiddleware := make([]negroni.Handler, len(baseMiddleware))
copy(apiMiddleware, baseMiddleware)
apiMiddleware = append(apiMiddleware,
	&ApiSpecificMiddleware{},
)

router.PathPrefix("/api/v1/users").Handler(negroni.New(append(
	apiMiddleware,
	&UsersSpecificMiddleware{},
	negroni.Wrap(usersRouter),
)...))
router.PathPrefix("/api/v1/posts").Handler(negroni.New(append(
	apiMiddleware,
	&PostsSpecificMiddleware{},
	negroni.Wrap(postsRouter),
)...))
router.PathPrefix("/api").Handler(negroni.New(append(
	apiMiddleware,
	negroni.Wrap(apiRouter),
)...))
```

If negroni had the `Append` functionality that Alice has, I believe my problems would be solved.

https://github.com/justinas/alice/blob/master/chain.go#L81

What I'm doing above with copying the array of handlers is, more or less, what Alice does with `Append`. 

Clearly, I can do it myself by managing my own array of `negroni.Handler` objects, but a convenience method in negroni like `Append` which could return a new negroni instance with the new handlers appended would probably be ideal for my case.

Then I could do something like this.

```
router.PathPrefix("/api/v1/users").Handler(negroniAPIMiddleware.Append(
	&UsersSpecificMiddleware{},
	negroni.Wrap(usersRouter),
))
router.PathPrefix("/api/v1/posts").Handler(negroniAPIMiddleware.Append(
	&PostsSpecificMiddleware{},
	negroni.Wrap(postsRouter),
))
router.PathPrefix("/api").Handler(negroniAPIMiddleware.Append(
	negroni.Wrap(apiRouter),
))
```

**Comment by jszwedko** [comment]
You could make it a little more concise via something like:

```golang
        // Negroni middleware that should be shared between multiple route groups.
        baseNegroniMiddleware := negroni.New(
                &LogMiddleware{},
                &BaseMiddleware1{},
                &BaseMiddleware2{},
        )

        router.PathPrefix("/api/v1/posts").Handler(negroni.New(append(
                baseNegroniMiddleware.Handlers(),
                &ApiSpecificMiddleware{},
                &UsersSpecificMiddleware{},
                negroni.Wrap(usersRouter),
        )...))
```

But I can definitely see the utility of of helper functions to extend a list of handlers. Possibly as a `.Append` on `Negroni` (similar to Alice) or perhaps as a top-level function like `negroni.Combine` or similar that would take `n` `Negroni` objects and create a new one chaining all of their handlers.

Open to PRs with proposals!

**Comment by williamhaley** [comment]
Thanks for the encouragement @yanfali and @jszwedko! 

Good call on the more concise usage @jszwedko. 

I've been a bit busy recently, but I'll try to take a stab at a PR for this enhancement sometime this week.

**Comment by williamhaley** [comment]
Closing this since my changes were merged in.
